lib/Sema/DerivedConformanceCodable.cpp

```diff
@@ -0,0 +1,949 @@
+    return false;
+
+  ASTContext &C = tc.Context;
+  auto parentDC = type->getDeclContext();
+  auto superType = type->getSuperclass();
+  return (bool)tc.conformsToProtocol(superType,
+                                     C.getProtocol(KnownProtocolKind::Codable),
+                                     parentDC,
+                                     ConformanceCheckFlags::Used);
+}
+
+// Looks up a type declaration with the given name in the Foundation module.
+//
+// Asserts that the Foundation module is loaded, that the given name does not
+// refer to more than one type, and that the entity with the given name refers
+// to a type instead of a different type of declaration.
+static TypeDecl* lookupFoundationTypeDecl(ASTContext &C, Identifier name) {
+  auto foundationModule = C.getLoadedModule(C.Id_Foundation);
+  assert(foundationModule && "Foundation module must be loaded.");
+
+  SmallVector<ValueDecl*, 1> results;
+  foundationModule->lookupMember(results, cast<DeclContext>(foundationModule),
+                                 DeclName(name), Identifier());
+  assert(results.size() == 1 && "Ambiguous/missing type.");
+
+  auto typeDecl = dyn_cast<TypeDecl>(results[0]);
+  assert(typeDecl && "Found non-type decl.");
+  return typeDecl;
+}
+
+// Looks up a type with the given name in the Foundation module.
+//
+// Asserts that the Foundation module is loaded, that the given name does not
+// refer to more than one type, and that the entity with the given name refers
+// to a type instead of a different type of declaration.
+static Type lookupFoundationType(ASTContext &C, Identifier name) {
+  return lookupFoundationTypeDecl(C, name)->getDeclaredInterfaceType();
+}
+
+// Validates that all the variables declared in the given list of declarations
+// conform to the Codable protocol.
+//
+// Produces a diagnostic on the given typechecker for every var which does not
+// conform. Calls a success callback for every var which does conform.
+template <typename ValidVarCallback>
+static bool
+validateVarsConformToCodable(TypeChecker &tc, DeclContext *context,
+                             NominalTypeDecl::StoredPropertyRange &vars,
+                             ValidVarCallback &callback) {
+  auto codableProto = tc.Context.getProtocol(KnownProtocolKind::Codable);
+  bool allConform = true;
+  for (auto varDecl : vars) {
+    // If the decl doesn't yet have a type, we may be seeing it before the type
+    // checker has gotten around to evaluating its type. For example:
+    //
+    // func foo() {
+    //   let b = Bar(from: decoder) // <- evaluates Bar conformance to Codable,
+    //                              //    forcing derivation
+    // }
+    //
+    // struct Bar : Codable {
+    //   var x: Int // <- we get to valuate x's var decl here, but its type
+    //              //    hasn't yet been evaluated
+    // }
+    //
+    // Validate the decl eagerly.
+    if (!varDecl->hasType())
+      tc.validateDecl(varDecl);
```

Comment:
I caught this case in testing exactly that — asking the `TypeChecker` to validate the decl eagerly evaluates the type so that past the `evaluateDecl` line, `hasType() == true`. However, if `getInterfaceType()` allows us to avoid doing this eager evaluation, then we should use that. Looking into it now.